text/0000-opentelemetry-protocol.md

```diff
@@ -0,0 +1,460 @@
+
+![Pipelined](images/otlp-pipelined.png)
+
+The number of unacknowledged requests for which the client did not receive responses yet is defined by the window size of underlying transport (for gRPC see details about [how the window size is calculated](https://grpc.io/blog/2017-08-22-grpc-go-perf-improvements/)).
+
+If the client detects that the underlying transport is broken and must be re-established the client MUST retry sending all requests that are still pending acknowledgement after re-establishing the transport. If the client is shutting down (e.g. when the containing process wants to exit) the client MUST wait until all pending acknowledgements are received or until an implementation specific timeout expires. This ensures reliable delivery of telemetry data.
+
+If the client is unable to deliver a certain request (e.g. it retried but failed multiple times or shutdown timer expired while waiting for acknowledgements) the client SHOULD record the fact that the data was not delivered.
+
+Pipelined mode is recommended when the client and the server are connected via a high-latency network but high throughput is still required. This is a typical situation when the client is an application running in one data center and the server is a telemetry backend running in a different data center, potentially in a completely different geography that is reachable via high-latency multi-hop network. Same recommendation applies to low latency networks if the server is capable of high processing rate but takes a long time to generate a response to each particular request.
+
+Note that the server does not need to know in what mode the client operates. The server just needs to ensure that it sends a response to every request. The Synchronized or Pipelined mode is enabled purely by client behavior.
+
+### Server Response
+
+The Server replies to every `Export` request received from the client by a corresponding `Export` response. The ID field of the response is set equal to the ID field of the request.
+
+#### Result Code
+
+`Export` response includes a `result_code` field, which indicates whether the server was able to successfully process the received data. Possible values for `result_code` field are:
+
+- `Success` - telemetry data is successfully processed by the server.
+
+- `FailedNoneRetryable` - processing of telemetry data failed. The client MUST NOT retry sending the same telemetry data. The telemetry data MUST be dropped. This for example can happen when the request contains bad data and cannot be deserialized or otherwise processed by the server. The client SHOULD maintain a counter of such dropped data.
+
+- `FailedRetryable` - processing of telemetry data failed. The client SHOULD record the error and may retry exporting the same data. This can happen when the server is temporarily unable to process the data.
+
+#### Throttling
```

Comment:
>My 2cents here, is that this is a hard problem to solve and I am not sure this is the right way. Let's assume you have 5 collectors, you need to have a global throttler to know when the other service are also overwhelmed or not.

You are right that generally it is a hard problem and there are no solutions that apply to all situations. 

However in some scenarios the "retry-after" approach can work well. For example when the same set of clients is connected to the same server and send data to the same server, the server can make approximate prediction about how overloaded it is (for example by looking at the length of its incoming queue that it needs to process and the rate of incoming data) and how much it needs the clients to slow down.

If the server is unable to calculate the throttle period, then it can use exponential backoff logic to calculate throttling period or use some fixed number.

This is analogous to Retry-After header in HTTP: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After and how the server can include it when returning status code 429 to ask client to slow down: https://tools.ietf.org/html/rfc6585#section-4

>Probably an easier solution to this is to return a FailedRetryable and implement some kind of backoff retry logic on the client. What do you think?

The problem with this approach is that it is not possible for client to understand the reason why the export request failed. I believe it is important for clients to distinguish general retryable errors from the case when the server is overloaded and asks the client to slow down. This is important for observability purposes (to know why exactly the export is failing).

If you think that having the numeric throttle period in the response is superfluous we can get rid of it and only report a boolean "overloaded" status in the response.